File: 2021/python/Day-03/Day_03.py
import time

with open("input.txt","r") as file:
    input_list = [line.strip() for line in file]

def solution01(input_list):
    gamma, epsilon = "",""
    length = len(input_list[0])

    for i in range(length):
        curr_bits = [bits[i] for bits in input_list]

        count_0 = curr_bits.count("0")
        count_1 = curr_bits.count("1")

        new_gamma_bit = "0" if count_0 > count_1 else "1"
        new_epsilon_bit = "0" if new_gamma_bit == "1" else "1"

        gamma += new_gamma_bit
        epsilon += new_epsilon_bit


    print(f"gamma: {gamma}")
    print(f"epsilon: {epsilon}")

    return int(gamma, 2) * int(epsilon, 2)

def solution02(input_list):

    def bit_criteria(bits, mode):
        bits_left = [i for i in bits]
    
        i = 0
        while len(bits_left) > 1:

            curr_bits = [bits[i] for bits in bits_left]

            count_0 = curr_bits.count("0")
            count_1 = curr_bits.count("1")

            if mode == "most":
                bit_to_keep = "1" if count_1 >= count_0 else "0"
            elif mode == "least":
                bit_to_keep = "0" if count_0 <= count_1 else "1"
            
            bits_left = [v for v in bits_left if v[i] == bit_to_keep]

            i += 1

        return bits_left[0]

    oxygen = bit_criteria(input_list, "most")
    co2 = bit_criteria(input_list, "least")

    print(f"oxygen: {oxygen}")
    print(f"co2: {co2}")

    return int(oxygen, 2) * int(co2, 2)


solution01_start_time = time.time()
solution01 = solution01(input_list)
solution01_total_time = time.time() - solution01_start_time
print(f"Solution 01: {solution01}, in {solution01_total_time}")

solution02_start_time = time.time()
solution02 = solution02(input_list)
solution02_total_time = time.time() - solution02_start_time
print(f"Solution 02: {solution02}, in {solution02_total_time}")

File: 2021/js/Day-03/Day-03.js
const fs = require("fs");

inputData = fs.readFileSync("input.txt", "utf8");

inputData = inputData.split(/\r?\n/);

// console.log(inputData);

const Solution01 = (inputData) => {
  const binaryLength = inputData[0].length;
  const reportLength = inputData.length;
  let [gamma, epsilon] = ["", ""];

  for (let pos = 0; pos < binaryLength; pos++) {
    let [ones, zeroes] = [0, 0];

    for (let line = 0; line < reportLength; line++) {
      if (inputData[line][pos] === "0") {
        zeroes += 1;
      } else {
        ones += 1;
      }
    }
    if (zeroes > ones) {
      gamma += 0;
      epsilon += 1;
    } else {
      gamma += 1;
      epsilon += 0;
    }
  }

  return parseInt(gamma, 2) * parseInt(epsilon, 2);
};

const Solution02 = (inputData) => {
  oxygen = [...inputData];
  co2 = [...inputData];
  const binaryLength = inputData[0].length;
  // const reportLength = inputData.length;
  // let [gamma, epsilon] = ["", ""];

  //find oxygen rating
  let pos = 0;
  while (oxygen.length > 1) {
    // for (let pos = 0; pos < binaryLength; pos++) {
    let [ones, zeroes] = [0, 0];

    //find most common
    for (let line = 0; line < oxygen.length; line++) {
      if (oxygen[line][pos] === "0") {
        zeroes += 1;
      } else {
        ones += 1;
      }
    }

    //keep only lines where most common bit is in current pos
    oxygen = oxygen.filter(
      (binary) => binary[pos] === (zeroes > ones ? "0" : "1")
    );

    // console.log(pos, zeroes, ones, oxygen);
    pos++;
  }

  //find co2 rating
  pos = 0;
  while (co2.length > 1) {
    // for (let pos = 0; pos < binaryLength; pos++) {
    let [ones, zeroes] = [0, 0];

    //find most common
    for (let line = 0; line < co2.length; line++) {
      if (co2[line][pos] === "0") {
        zeroes += 1;
      } else {
        ones += 1;
      }
    }

    //keep only lines where least common bit is in current pos
    co2 = co2.filter((binary) => binary[pos] === (zeroes > ones ? "1" : "0"));
    // console.log(pos, zeroes, ones, co2);
    pos++;
  }

  return parseInt(oxygen[0], 2) * parseInt(co2[0], 2);
};

console.log(`Solution 01: ${Solution01(inputData)}`);
console.log(`Solution 02: ${Solution02(inputData)}`);
